From: Ray Strode <rstrode@redhat.com>
Date: Tue, 6 Feb 2024 13:06:32 -0500
Subject: authPrompt: Parameterize reset function

In the future, the code will need to do a partial reset where
some state is carried over or explicitly specified.

This commit prepares for that by allowing the request type and
querying service to be specified at reset time.
---
 js/gdm/authPrompt.js | 45 +++++++++++++++++++++++++--------------------
 1 file changed, 25 insertions(+), 20 deletions(-)

diff --git a/js/gdm/authPrompt.js b/js/gdm/authPrompt.js
index 8a35efe..1ee0708 100644
--- a/js/gdm/authPrompt.js
+++ b/js/gdm/authPrompt.js
@@ -619,7 +619,12 @@ export const AuthPrompt = GObject.registerClass({
             this._updateEntry(false);
     }
 
-    reset() {
+    reset(params) {
+        let {beginRequestType, queryingService} = Params.parse(params, {
+            beginRequestType: null,
+            queryingService: null,
+        });
+
         let oldStatus = this.verificationStatus;
         this.verificationStatus = AuthPromptStatus.NOT_VERIFYING;
         this.cancelButton.reactive = this._hasCancelButton;
@@ -629,7 +634,7 @@ export const AuthPrompt = GObject.registerClass({
         if (this._userVerifier)
             this._userVerifier.cancel();
 
-        this._queryingService = null;
+        this._queryingService = queryingService;
         this.clear();
         this._message.opacity = 0;
         this.setUser(null);
@@ -641,24 +646,24 @@ export const AuthPrompt = GObject.registerClass({
         else if (oldStatus === AuthPromptStatus.VERIFICATION_CANCELLED)
             this.emit('cancelled');
 
-        let beginRequestType;
-
-        if (this._mode === AuthPromptMode.UNLOCK_ONLY) {
-            // The user is constant at the unlock screen, so it will immediately
-            // respond to the request with the username
-            if (oldStatus === AuthPromptStatus.VERIFICATION_CANCELLED)
-                return;
-            beginRequestType = BeginRequestType.PROVIDE_USERNAME;
-        } else if (this._userVerifier.foregroundServiceDeterminesUsername()) {
-            // We don't need to know the username if the user preempted the login screen
-            // with a smartcard or with preauthenticated oVirt credentials
-            beginRequestType = BeginRequestType.DONT_PROVIDE_USERNAME;
-        } else if (oldStatus === AuthPromptStatus.VERIFICATION_IN_PROGRESS) {
-            // We're going back to retry with current user
-            beginRequestType = BeginRequestType.REUSE_USERNAME;
-        } else {
-            // In all other cases, we should get the username up front.
-            beginRequestType = BeginRequestType.PROVIDE_USERNAME;
+        if (beginRequestType === null) {
+            if (this._mode === AuthPromptMode.UNLOCK_ONLY) {
+                // The user is constant at the unlock screen, so it will immediately
+                // respond to the request with the username
+                if (oldStatus === AuthPromptStatus.VERIFICATION_CANCELLED)
+                    return;
+                beginRequestType = BeginRequestType.PROVIDE_USERNAME;
+            } else if (this._userVerifier.foregroundServiceDeterminesUsername()) {
+                // We don't need to know the username if the user preempted the login screen
+                // with a smartcard or with preauthenticated oVirt credentials
+                beginRequestType = BeginRequestType.DONT_PROVIDE_USERNAME;
+            } else if (oldStatus === AuthPromptStatus.VERIFICATION_IN_PROGRESS) {
+                // We're going back to retry with current user
+                beginRequestType = BeginRequestType.REUSE_USERNAME;
+            } else {
+                // In all other cases, we should get the username up front.
+                beginRequestType = BeginRequestType.PROVIDE_USERNAME;
+            }
         }
 
         this.emit('reset', beginRequestType);
